drivers/include/periph/gpio_ll_irq.h

```diff
@@ -0,0 +1,180 @@
+/**
+ * @brief   Definition of possible IRQ triggers
+ */
+typedef enum {
+    GPIO_TRIGGER_EDGE_FALLING,  /**< edge triggered IRQ on falling flanks only
+                                  */
+    GPIO_TRIGGER_EDGE_RISING,   /**< edge triggered IRQ on rising flanks only */
+    GPIO_TRIGGER_EDGE_BOTH,     /**< edge triggered IRQ on falling *AND* rising
+                                     flanks */
+    GPIO_TRIGGER_LEVEL_HIGH,    /**< level triggered IRQ on high input */
+    GPIO_TRIGGER_LEVEL_LOW      /**< level triggered IRQ on low input */
+} gpio_irq_trig_t;
+#endif
+
+/**
+ * @brief   Signature of event callback function that is called on IRQs
+ * @param[in] arg       optional context for the callback
+ * @warning This function is called in IRQ context
+ */
+typedef void (*gpio_ll_cb_t)(void *arg);
+
+/**
+ * @brief   Set up an IRQ for the given GPIO pin and activate it
+ *
+ * @param[in] port      port the pin belongs to
+ * @param[in] pin       number of the pin to setup IRQs on
+ * @param[in] trig      define what triggers the IRQ
+ * @param[in] cb        callback that is called from interrupt context
+ * @param[in] arg       argument to be passed to the callback
+ *
+ * @pre     The given GPIO pin has been initialized using
+ *          @ref gpio_ll_init prior to this call
+ *
+ * @retval  0           Success
+ * @retval  -ENOTSUP    Either requested trigger not supported on the specified
+ *                      pin, or no external IRQs supported on that pin at all
+ * @retval  -EBUSY      All hardware entities that monitor GPIOs and issue IRQs
+ *                      on the requested trigger are busy monitoring other
+ *                      GPIOs. Call @ref gpio_ll_irq_off on one of those and
+ *                      retry.
+ */
+int gpio_ll_irq(gpio_port_t port, uint8_t pin, gpio_irq_trig_t trig,
+                gpio_ll_cb_t cb, void *arg);
+
+/**
+ * @brief   Mask IRQs on the given GPIO pin
+ *
+ * @param[in] port      port the pin belongs to
+ * @param[in] pin       number of the pin to disable IRQs on
+ *
+ * @note    IRQs can be re-enabled without reconfiguration using
+ *          @ref gpio_ll_irq_enable
+ *
+ * If IRQs are not needed for a longer term, consider using @ref
+ * gpio_ll_irq_off instead.
+ */
+void gpio_ll_irq_mask(gpio_port_t port, uint8_t pin);
+
+#if HAS_GPIO_LL_IRQ_UNMASK || defined(DOXYGEN)
+/**
+ * @brief   Unmask IRQs on the given GPIO pin
+ *
+ * Same as @ref gpio_ll_irq_unmask_clear except that IRQs that came in during
+ *
+ * @warning On same MCUs (most notably STM32) this is impossible to implement.
+ *          Check @ref HAS_GPIO_LL_IRQ_UNMASK to see if this function is
+ *          available, or (usually better) just use
+ *          @ref gpio_ll_irq_unmask_clear as portable alternative.
+ *
+ * @param[in] port      port the pin belongs to
+ * @param[in] pin       number of the pin to unmask IRQs on
+ */
+void gpio_ll_irq_unmask(gpio_port_t port, uint8_t pin);
```

Comment:
Why do we define these function bit-wise instead of port-wise.

Reply:
Wouldn't it be a nice feature to disable or enable the interrupt for a number of pins in one function call?

Reply:
OK, on ESP32 they can't be enabled portwise but only bitwise.

Reply:
IRQ stuff was not possible to implement port-wise for all platforms and I couldn't find any use case for configuring IRQs port-wise instead if pin-wise.